src/rad/mode.clj

```diff
@@ -0,0 +1,57 @@
+(ns rad.mode
+  "State and functions for entering/leaving modes such as :insert and :command"
+  (:require [rad.buffer]
+            [rad.terminal]))
+
+;; This works only in repl right now ;)
+
+;; A rad command is a function. If it returns another function, it will not evaluate. If you give it something which is not fn?, it will evaluate.
+;; To make commands which take arguments - they return a function that takes a function. This creates a chain, and that chain is the arguments.
+
+(def current-mode (atom :command))           ; nil deafult to insert mode ; FIXME
+
+(defn rep
+  "Runs f n times"
+  [n f & acc]
+  (if (< 0 n)
+    (recur (dec n) f (conj acc (f)))
+    acc))
+
+(defn r []                              ;make me a macro
+  (fn [n]
+    (fn [f]
+      (rep n f))))
+
+(defn key-map [] {\r r
```

Comment:
Map the \newline or :enter keys to return `nil`. That will act as a "end command"-command. 

Then you can use it to cancel any command, but also to say "input is done". For instance, there can be a `d` command, which can take "direction". Direction can be `word`, `sentence`, `closing bracket` or `backwards` (etc). But if you hit enter instead, it will get a nil and just run it's default (which is `buffer/delete-char!` or something)